cnmodel/data/_db.py

```diff
@@ -208,13 +213,30 @@ def add_table_data(name, row_key, col_key, data, **kwds):
     #print col_names
     #print row_names
     #print cells
+
+    changes = [] # a list of parameters that are changed if we are rewriting a table
     for i,row in enumerate(row_names):
         for j,col in enumerate(col_names):
             kwds[row_key] = row
             kwds[col_key] = col
-            setval(cells[i][j], name, **kwds)
+            oldval = setval(cells[i][j], name, **kwds)
+            if oldval is not None and oldval != cells[i][j]:
+                key = mk_key(name, **kwds)
+                changes.append({'key': key, 'new': cells[i][j], 'old': oldval, 'name': name})
```

Comment:
The name bit is redundant since it gets included in the key, but it should work as written.